README.md
# Homebrew Libraries Kit - Timers

## Known issues

- Some tests may fail due to accidental awakening of *condition_variable*

## License

<img align="right" src="https://www.gnu.org/graphics/lgplv3-with-text-154x68.png">

The library is licensed under [GNU Lesser General Public License 3.0](https://www.gnu.org/licenses/lgpl-3.0.txt):

Copyright © 2021 Dmitry Plastinin

Hlk Timers is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as pubblished by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

Hlk Timers is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser Public License for more details

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 2.1.3 - 2021-10-19

### Fixed

- Segmentation fault...again.

## 2.1.2 - 2021-10-14

### Fixed

- Perfomance and stability improvements

## 2.1.1 - 2021-10-14

### Changed

- Timer signal

## 2.1.0 - 2021-10-11

### Added

- Dispatcher thread: now the signal wakes up the dispatcher thread instead of creating the job

### Fixed

- Signal handler used non signal-safety functions

## 2.0.1 - 2021-10-09

### Fixed

- Threads signal mask

## 2.0.0 - 2021-09-28

### Changed

- Timers are now based on real time signal rather than file descriptors so the number of timers is now unlimited

## 1.2.6 - 2021-09-24

### Fixed

- Skip second when nanoseconds were rounded

## 1.2.5 - 2021-09-22

### Fixed

- Deadlock when the timer was started and the one-time start flag was checked 

## 1.2.4 - 2021-09-21

### Fixed

- Bug when the timer descriptor was reset to -1 in the timer handler thread, but another thread updated the timer 

## 1.2.3 - 2021-09-03

### Fixed

- Bug when timer was deleted after stopping itself onTimeout() handler and starting again

## 1.2.2 - 2021-09-03

### Added

- Change log

### Fixed

- Fixed handling of file descriptor zero

tests/stoptimer.cpp
#include "hlk/timers/timer.h"

#include <fcntl.h>
#include <iostream>
#include <unistd.h>
#include <signal.h>
#include <execinfo.h>
#include <condition_variable>

std::condition_variable cv;
bool triggered = false;
Hlk::Timer timer;

void sigsegvHandler(int signum) {
    int nptrs, fd;
    void *buffer[1024];

    nptrs = backtrace(buffer, 1024);
    fd = open("cascadetimercall_backtrace.txt", O_CREAT | O_WRONLY | O_TRUNC, 0665);
    backtrace_symbols_fd(buffer, nptrs, fd);
    close(fd);

    signal(signum, SIG_DFL);
    exit(3);
}

void timerTimeoutHandler() {
    std::cout << "Timer triggered\n";
    triggered = true;
    cv.notify_one();
}

int main(int argc, char *argv[]) {
    signal(SIGSEGV, sigsegvHandler);
    
    timer.onTimeout.addEventHandler(timerTimeoutHandler);
    timer.start(100);
    timer.stop();

    std::mutex mutex;
    std::unique_lock lock(mutex);
    cv.wait_for(lock, std::chrono::seconds(1), [] () { return triggered; });
    if (triggered) {
        std::cout << "Test not passed\n";
        return EXIT_FAILURE;
    }

    std::cout << "Test passed\n";
    return EXIT_SUCCESS;
}

tests/concurrence.cpp
#include "hlk/timers/timer.h"

#include <fcntl.h>
#include <iostream>
#include <unistd.h>
#include <signal.h>
#include <execinfo.h>
#include <condition_variable>

bool threadsRunning = true;
unsigned int triggerCount = 0;
unsigned int startCount = 0;
unsigned int stopCount = 0;

void sigsegvHandler(int signum) {
    int nptrs, fd;
    void *buffer[1024];

    nptrs = backtrace(buffer, 1024);
    fd = open("cascadetimercall_backtrace.txt", O_CREAT | O_WRONLY | O_TRUNC, 0665);
    backtrace_symbols_fd(buffer, nptrs, fd);
    close(fd);

    signal(signum, SIG_DFL);
    exit(3);
}

void timerTimeoutHandler() {
    ++triggerCount;
}

int main(int argc, char *argv[]) {
    signal(SIGSEGV, sigsegvHandler);

    Hlk::Timer timer;
    timer.onTimeout.addEventHandler(timerTimeoutHandler);

    std::thread thread1([&timer] () {
        while (threadsRunning) {
            timer.stop();
            ++stopCount;
        }
    });

    std::thread thread2([&timer] () {
        while (threadsRunning) {
            timer.start(1);
            ++startCount;
        }
    });

    std::thread thread3([&timer] () {
        while (threadsRunning) {
            timer.stop();
            ++stopCount;
        }
    });

    std::thread thread4([&timer] () {
        while (threadsRunning) {
            timer.start(0);
            ++startCount;
        }
    });

    sleep(10);

    threadsRunning = false;
    thread1.join();
    thread2.join();
    thread3.join();
    thread4.join();

    std::cout << "Test passed!\n";
    std::cout << "The timer triggered: " << triggerCount << " times\n";
    std::cout << "The timer has been started: " << startCount << " times\n";
    std::cout << "The timer has been stopped: " << stopCount << " times\n";

    return 0;
}

tests/stopfromtimer.cpp
#include "hlk/timers/timer.h"

#include <fcntl.h>
#include <iostream>
#include <unistd.h>
#include <signal.h>
#include <execinfo.h>
#include <condition_variable>

std::condition_variable cv;
bool triggered = false;
Hlk::Timer timer;

void sigsegvHandler(int signum) {
    int nptrs, fd;
    void *buffer[1024];

    nptrs = backtrace(buffer, 1024);
    fd = open("cascadetimercall_backtrace.txt", O_CREAT | O_WRONLY | O_TRUNC, 0665);
    backtrace_symbols_fd(buffer, nptrs, fd);
    close(fd);

    signal(signum, SIG_DFL);
    exit(3);
}

void timerTimeoutHandler() {
    std::cout << "Timer triggered\n";
    timer.stop();
    triggered = true;
    cv.notify_one();
}

int main(int argc, char *argv[]) {
    signal(SIGSEGV, sigsegvHandler);
    
    timer.onTimeout.addEventHandler(timerTimeoutHandler);
    timer.start(1);

    std::mutex mutex;
    std::unique_lock lock(mutex);
    cv.wait_for(lock, std::chrono::seconds(3), [] () { return triggered; });
    if (!triggered) {
        std::cout << "Test not passed\n";
        return EXIT_FAILURE;
    }

    std::cout << "Test passed\n";
    return EXIT_SUCCESS;
}

include/hlk/timers/timer.h
/******************************************************************************
 * 
 * Copyright (C) 2021 Dmitry Plastinin
 * Contact: uncellon@yandex.ru, uncellon@gmail.com, uncellon@mail.ru
 * 
 * This file is part of the Hlk Timers library.
 * 
 * Hlk Timers is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as pubblished by the
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * Hlk Timers is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser Public License for more
 * details
 * 
 * You should have received a copy of the GNU Lesset General Public License
 * along with Hlk Timers. If not, see <https://www.gnu.org/licenses/>.
 * 
 *****************************************************************************/

#ifndef HLK_TIMER_H
#define HLK_TIMER_H

#include <thread>
#include <hlk/events/event.h>

namespace Hlk {

class Timer {
public:
    /**************************************************************************
     * Constructors / Destructors
     *************************************************************************/

    Timer();
    ~Timer();

    /**************************************************************************
     * Public methods
     *************************************************************************/

    void start(unsigned int msec);
    void stop();

    /**************************************************************************
     * Events
     *************************************************************************/

    Event<> onTimeout;

    /**************************************************************************
     * Accessors / Mutators
     *************************************************************************/

    bool oneShot() const;
    void setOneShot(bool value);

    bool started() const;

protected:
    /**************************************************************************
     * Static: Methods
     *************************************************************************/

    static int reserveId(Timer *timer);
    static void dispatcherLoop();

    /**************************************************************************
     * Static: Members
     *************************************************************************/

    static std::mutex m_cdtorMutex;
    static std::mutex m_timerMutex;
    static unsigned int m_counter;
    static std::thread *m_dispatcherThread;
    static bool m_dispatcherRunning;
    static std::vector<Timer *> m_timerInstances;

    /**************************************************************************
     * Protected: Methods
     *************************************************************************/

    void createTimer(unsigned int msec);
    void deleteTimer();
    void setTime(unsigned int msec);

    /**************************************************************************
     * Protected: Members
     *************************************************************************/

    bool m_oneShot = false;
    bool m_started = false;
    int m_index = -1;
    timer_t m_timerid = timer_t();
};

/******************************************************************************
 * Inline definition: Accessors / Mutators
 *****************************************************************************/

inline bool Timer::started() const { return m_started; }

inline bool Timer::oneShot() const { return m_oneShot; }
inline void Timer::setOneShot(bool value) { m_oneShot = value; }

} // namespace Hlk

#endif // HLK_TIMER_H

include/hlk/timers/timermanager.h
/******************************************************************************
 * 
 * Copyright (C) 2021 Dmitry Plastinin
 * Contact: uncellon@yandex.ru, uncellon@gmail.com, uncellon@mail.ru
 * 
 * This file is part of the Hlk Timers library.
 * 
 * Hlk Timers is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as pubblished by the
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * Hlk Timers is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser Public License for more
 * details
 * 
 * You should have received a copy of the GNU Lesset General Public License
 * along with Hlk Timers. If not, see <https://www.gnu.org/licenses/>.
 * 
 *****************************************************************************/

#ifndef HLK_TIMER_MANAGER_H
#define HLK_TIMER_MANAGER_H

#include <poll.h>
#include <vector>
#include <thread>
#include <hlk/events/event.h>

namespace Hlk {

/******************************************************************************
 * Main class
 *****************************************************************************/

class TimerManager {
public:
    /**************************************************************************
     * Constructors / Destructors
     *************************************************************************/

    TimerManager();
    ~TimerManager();

    /**************************************************************************
     * Public methods
     *************************************************************************/

    int createTimer(unsigned int msec, bool oneShot, Hlk::Delegate<void> callback);
    void deleteTimer(int timerfd);
    void updateTimer(int timerfd, unsigned int msec, bool oneShot);

protected:
    /**************************************************************************
     * Private methods
     *************************************************************************/

    void loop();
    void writeSafeInterrupt();

    /**************************************************************************
     * Private members
     *************************************************************************/

    std::vector<pollfd> m_pfds;
    std::vector<Hlk::Delegate<void>> m_callbacks;

    std::mutex m_pfdsMutex;
    std::mutex m_readWriteMutex;

    std::thread *m_thread = nullptr;
    bool m_running = false;
    int m_pipes[2];
    int m_bytes = 0;
};

} // namespace Hlk

#endif // HLK_TIMER_MANAGER_H

include/hlk/timers/timermanager.cpp
/******************************************************************************
 * 
 * Copyright (C) 2021 Dmitry Plastinin
 * Contact: uncellon@yandex.ru, uncellon@gmail.com, uncellon@mail.ru
 * 
 * This file is part of the Hlk Timers library.
 * 
 * Hlk Timers is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as pubblished by the
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * Hlk Timers is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser Public License for more
 * details
 * 
 * You should have received a copy of the GNU Lesset General Public License
 * along with Hlk Timers. If not, see <https://www.gnu.org/licenses/>.
 * 
 *****************************************************************************/

#include "timermanager.h"

#include <fcntl.h>
#include <cstring>
#include <unistd.h>
#include <stdexcept>
#include <sys/timerfd.h>

namespace Hlk {

/******************************************************************************
 * Constructors / Destructors
 *****************************************************************************/

TimerManager::TimerManager() {
    // Create a pipe to be able to interrupt the timer thread
    if (pipe(m_pipes) == -1) {
        throw std::runtime_error("pipe(...) failed");
    }

    // Create pipe polling descriptor
    pollfd pfd;
    pfd.fd = m_pipes[0];
    pfd.events = POLLIN;
    m_pfds.push_back(pfd);

    // Create thread
    m_running = true;
    m_thread = new std::thread(&TimerManager::loop, this);
}

TimerManager::~TimerManager() {
    // Delete thread
    m_running = false;
    m_readWriteMutex.lock();
    writeSafeInterrupt();
    m_readWriteMutex.unlock();
    m_thread->join();
    delete m_thread;

    // Close timers
    for (size_t i = 0; i < m_pfds.size(); ++i) {
        close(m_pfds[i].fd);
    }
    m_pfds.clear();
    m_callbacks.clear();

    // Close pipes
    close(m_pipes[0]);
    close(m_pipes[1]);
}

/******************************************************************************
 * Public methods
 *****************************************************************************/

int TimerManager::createTimer(unsigned int msec, bool oneShot, Hlk::Delegate<void> callback) {
    // Create timerfd
    int timerfd = timerfd_create(CLOCK_MONOTONIC, 0);
    if (timerfd == -1) {
        if (errno == EMFILE) {
            throw std::runtime_error("timerfd_create(...) failed, too many timers");
        }
        throw std::runtime_error("timerfd_create(...) failed");
    }

    // Get current time
    struct timespec tm;
    clock_gettime(CLOCK_MONOTONIC, &tm);

    // Set new timer values
    itimerspec spec;
    spec.it_value.tv_sec = tm.tv_sec + (msec / 1000);
    spec.it_value.tv_nsec = (tm.tv_nsec + (msec % 1000) * 1000000) % 1000000000;
    if (oneShot) {
        spec.it_interval.tv_sec = 0;
        spec.it_interval.tv_nsec = 0;
    } else {
        spec.it_interval.tv_sec = msec / 1000;
        spec.it_interval.tv_nsec = (msec % 1000) * 1000000;
    }

    // Apply new timer values
    if (timerfd_settime(timerfd, TFD_TIMER_ABSTIME, &spec, nullptr) == -1) {
        throw std::runtime_error("timerfd_settime(...) failed");
    }

    // Create polling descriptor
    pollfd pfd;
    pfd.fd = timerfd;
    pfd.events = POLLIN;

    // Interrupt thread
    m_readWriteMutex.lock();
    if (!m_pfdsMutex.try_lock()) {
        writeSafeInterrupt();
        m_pfdsMutex.lock();
    }

    // Append data
    m_callbacks.push_back(callback);
    m_pfds.push_back(pfd);

    // Resume thread
    m_pfdsMutex.unlock();
    m_readWriteMutex.unlock();

    return pfd.fd;
}

void TimerManager::deleteTimer(int timerfd) {
    // Interrupt thread
    m_readWriteMutex.lock();
    if (!m_pfdsMutex.try_lock()) {
        writeSafeInterrupt();
        m_pfdsMutex.lock();
    }

    // Delete timer data
    for (size_t i = 1; i < m_pfds.size(); ++i) {
        if (m_pfds[i].fd != timerfd) continue;
        close(m_pfds[i].fd);
        m_pfds[i].fd = 0;
        break;
    }

    // Resume thread
    m_pfdsMutex.unlock();
    m_readWriteMutex.unlock();
}

void TimerManager::updateTimer(int timerfd, unsigned int msec, bool oneShot) {
    // Get current time
    struct timespec tm;
    clock_gettime(CLOCK_MONOTONIC, &tm);

    // Set new timer values
    itimerspec timerspec;
    memset(&timerspec, 0, sizeof(itimerspec));
    timerspec.it_value.tv_sec = tm.tv_sec + (msec / 1000);
    timerspec.it_value.tv_nsec = (tm.tv_nsec + (msec % 1000) * 1000000) % 1000000000;
    if (oneShot) {
        timerspec.it_interval.tv_sec = 0;
        timerspec.it_interval.tv_nsec = 0;
    } else {
        timerspec.it_interval.tv_sec = msec / 1000;
        timerspec.it_interval.tv_nsec = (msec % 1000) * 1000000;
    }

    // Apply new timer values
    if (timerfd_settime(timerfd, TFD_TIMER_ABSTIME, &timerspec, nullptr) == -1) {
        throw std::runtime_error("timerfd_settime(...) failed");
    }

    // Interrupt thread
    m_readWriteMutex.lock();
    writeSafeInterrupt();
    m_readWriteMutex.unlock();
}

/******************************************************************************
 * Private methods
 *****************************************************************************/

void TimerManager::loop() {
    int ret = 0;
    ssize_t bytesRead = 0;
    uint64_t exp = 0;
    size_t i = 1;

    while (m_running) {
        m_pfdsMutex.lock();
        ret = poll(m_pfds.data(), m_pfds.size(), -1);
        m_pfdsMutex.unlock();

        if (ret <= 0) continue;

        // Clear pipe
        if (m_pfds[0].revents == POLLIN) {
            m_pfds[0].revents = 0;

            m_readWriteMutex.lock();
            do {
                bytesRead = read(m_pfds[0].fd, &exp, sizeof(uint64_t));
                m_bytes -= bytesRead;
            } while (bytesRead == sizeof(uint64_t));
            m_readWriteMutex.unlock();

            if (ret == 1) continue;
        }

        // Process timers
        m_pfdsMutex.lock();
        for (i = 1; i < m_pfds.size(); ++i) {
            if (m_pfds[i].revents != POLLIN) continue;

            m_pfds[i].revents = 0;
            bytesRead = read(m_pfds[i].fd, &exp, sizeof(uint64_t));    

            if (bytesRead != sizeof(uint64_t)) continue;

            // Timer was deleted
            if (m_pfds[i].fd == 0) {
                m_pfds.erase(m_pfds.begin() + i);
                m_callbacks.erase(m_callbacks.begin() + i - 1);
                --i;
                continue;
            }

            m_pfdsMutex.unlock();
            m_callbacks[i - 1]();
            m_pfdsMutex.lock();
        }
        m_pfdsMutex.unlock();
    }
}

inline void TimerManager::writeSafeInterrupt() {
    // Write two bytes to avoid lock by read(...)
    if ((m_bytes + 1) % sizeof(uint64_t) == 0) {
        write(m_pipes[1], reinterpret_cast<const void *>("00"), 2);
        m_bytes += 2;
    } else {
        write(m_pipes[1], reinterpret_cast<const void *>("0"), 1);
        ++m_bytes;
    }
}

} // namespace Hlk

include/hlk/timers/timer.cpp
/******************************************************************************
 * 
 * Copyright (C) 2021 Dmitry Plastinin
 * Contact: uncellon@yandex.ru, uncellon@gmail.com, uncellon@mail.ru
 * 
 * This file is part of the Hlk Timers library.
 * 
 * Hlk Timers is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as pubblished by the
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * Hlk Timers is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser Public License for more
 * details
 * 
 * You should have received a copy of the GNU Lesset General Public License
 * along with Hlk Timers. If not, see <https://www.gnu.org/licenses/>.
 * 
 *****************************************************************************/

#include "timer.h"

#include <vector>
#include <unistd.h>
#include <signal.h>
#include <ucontext.h>
#include <hlk/pool/pool.h>

#define TIMER_SIGNAL SIGRTMIN

namespace Hlk {

/******************************************************************************
 * Static members
 *****************************************************************************/

std::mutex Timer::m_cdtorMutex;
std::mutex Timer::m_timerMutex;
unsigned int Timer::m_counter = 0;
std::thread *Timer::m_dispatcherThread = nullptr;
bool Timer::m_dispatcherRunning = false;
std::vector<Timer *> Timer::m_timerInstances;

/******************************************************************************
 * Constructors / Destructors
 *****************************************************************************/

static void timerSignalHandler(int sig, siginfo_t *info, void *ucontext) {
    sigaddset(&static_cast<ucontext_t *>(ucontext)->uc_sigmask, TIMER_SIGNAL);
    kill(getpid(), TIMER_SIGNAL);
}

Timer::Timer() {
    std::unique_lock lock(m_cdtorMutex);

    if (++m_counter != 1) {
        return;
    }

    m_timerInstances.clear();
    
    sigset_t sigset;
    sigemptyset(&sigset);
    sigaddset(&sigset, TIMER_SIGNAL);
    sigprocmask(SIG_BLOCK, &sigset, nullptr);

    struct sigaction sa;
    sa.sa_flags = SA_SIGINFO | SA_ONSTACK;
    sa.sa_sigaction = timerSignalHandler;
    sigaction(TIMER_SIGNAL, &sa, nullptr);
    kill(getpid(), TIMER_SIGNAL);

    m_dispatcherRunning = true;
    m_dispatcherThread = new std::thread(&Timer::dispatcherLoop);
}

Timer::~Timer() {
    stop();

    std::unique_lock lock(m_cdtorMutex);

    if (--m_counter != 0) {
        return;
    }

    m_dispatcherRunning = false;
    sigqueue(getpid(), TIMER_SIGNAL, sigval());
    m_dispatcherThread->join();
    delete m_dispatcherThread;
    m_dispatcherThread = nullptr;

    m_timerInstances.clear();
}

/******************************************************************************
 * Public methods
 *****************************************************************************/

void Timer::start(unsigned int msec) {
    std::unique_lock lock(m_timerMutex);
    if (!m_started) {
        m_started = true;
        createTimer(msec);
    }
    setTime(msec);
}

void Timer::stop() {
    std::unique_lock lock(m_timerMutex);

    if (!m_started) {
        return;
    }

    m_started = false;
    deleteTimer();
    m_timerid = timer_t();
    m_index = -1;
}

/******************************************************************************
 * Static: Methods
 *****************************************************************************/

inline int Timer::reserveId(Timer *timer) {
    for (size_t i = 0; i < m_timerInstances.size(); ++i) {
        if (m_timerInstances[i] != nullptr) {
            continue;
        }
        m_timerInstances[i] = timer;
        return i;
    }

    m_timerInstances.push_back(timer);
    return m_timerInstances.size() - 1;
}

void Timer::dispatcherLoop() {
    Hlk::Pool *poolInstance =Pool::getInstance();
    Hlk::Timer *timerInstance= nullptr;
    int index = 0;

    sigset_t sigset;
    sigemptyset(&sigset);
    sigaddset(&sigset, TIMER_SIGNAL);

    siginfo_t siginfo;

    while (m_dispatcherRunning) {
        sigwaitinfo(&sigset, &siginfo);
        
        if (siginfo.si_code != SI_TIMER) {
            continue;
        }

        std::unique_lock lock(m_timerMutex);

        index = siginfo._sifields._timer.si_sigval.sival_int;
        timerInstance = m_timerInstances[index];
        if (!timerInstance || !timerInstance->m_started) {
            continue;
        }
        poolInstance->pushTask([timerInstance] () {
            timerInstance->onTimeout();
        });
        if (timerInstance->oneShot()) {
            timerInstance->m_started = false;
            timerInstance->deleteTimer();
        }
    }
}

/******************************************************************************
 * Protected: Methods
 *****************************************************************************/

inline void Timer::createTimer(unsigned int msec) {
    // Reserve signal for this object
    m_index = reserveId(this);

    // Create sigevent for timer
    struct sigevent sev;
    sev.sigev_notify = SIGEV_SIGNAL;
    sev.sigev_signo = TIMER_SIGNAL;
    sev.sigev_value.sival_ptr = &m_timerid;
    sev.sigev_value.sival_int = m_index;
    sev._sigev_un._sigev_thread._attribute = nullptr;

    if (timer_create(CLOCK_MONOTONIC, &sev, &m_timerid) == -1) {
        throw std::runtime_error("timer_create(...) failed, errno: "
            + std::to_string(errno));
    }
}

inline void Timer::deleteTimer() {
    timer_delete(m_timerid);
    m_timerInstances[m_index] = nullptr;
}

inline void Timer::setTime(unsigned int msec) {
    timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);

    itimerspec its;
    auto nsec = (msec % 1000) * 1000000;
    its.it_value.tv_sec = ts.tv_sec + (msec / 1000) + (ts.tv_nsec + nsec) / 1000000000;
    its.it_value.tv_nsec = (ts.tv_nsec + nsec) % 1000000000;
    if (m_oneShot) {
        its.it_interval.tv_sec = 0;
        its.it_interval.tv_nsec = 0;
    } else {
        its.it_interval.tv_sec = msec / 1000;
        its.it_interval.tv_nsec = (msec % 1000) * 1000000;
    }

    if (timer_settime(m_timerid, TIMER_ABSTIME, &its, nullptr) == -1) {
        throw std::runtime_error("timerfd_settime(...) failed, errno: " 
            + std::to_string(errno));
    }
}

} // namespace Hlk